man/lib/c/malloc.man
.\" Copyright (c) 1980 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)malloc.3	6.3 (Berkeley) 5/14/86
.\"
.TH MALLOC 3  "May 14, 1986"
.UC 4
.SH NAME
malloc, free, realloc, calloc, alloca \- memory allocator
.SH SYNOPSIS
.nf
#include <stdlib.h>
.PP
.B char *malloc(size)
.B unsigned size;
.PP
.B free(ptr)
.B char *ptr;
.PP
.B char *realloc(ptr, size)
.B char *ptr;
.B unsigned size;
.PP
.B char *calloc(nelem, elsize)
.B unsigned nelem, elsize;
.PP
.B char *alloca(size)
.B int size;
.fi
.SH DESCRIPTION
.I Malloc
and
.I free
provide a general-purpose memory allocation package.
.I Malloc
returns a pointer to a block of at least
.I size
bytes beginning on a word boundary.
.PP
The argument to
.I free
is a pointer to a block previously allocated by
.IR malloc ;
this space is made available for further allocation,
but its contents are left undisturbed.
.PP
Needless to say, grave disorder will result if the space assigned by
.I malloc
is overrun or if some random number is handed to
.IR free .
.PP
.I Malloc
maintains multiple lists of free blocks according to size,
allocating space from the appropriate list.
It calls
.I sbrk
(see
.IR brk (2))
to get more memory from the system when there is no
suitable space already free.
.PP
.I Realloc
changes the size of the block pointed to by
.I ptr
to
.I size
bytes and returns a pointer to the (possibly moved) block.
The contents will be unchanged up to the lesser of the new and old sizes.
The Sprite version of \fIrealloc\fR differs from other UNIX versions
in that \fIptr\fR may not point to a free block (in other UNIX
versions \fIptr\fr may poin to a block freed since the last call to
\fImalloc\fR, \fIcalloc\fR, or \fIrealloc\fR).
.PP
.I Calloc
allocates space for an array of
.I nelem
elements of size
.I elsize.
The space is initialized to zeros.
.PP
.I Alloca
allocates 
.I size
bytes of space in the stack frame of the caller.
This temporary space is automatically freed on
return.
.PP
Each of the allocation routines returns a pointer
to space suitably aligned (after possible pointer coercion)
for storage of any type of object.
.SH SEE ALSO
brk(2),
pagesize(2)
.SH DIAGNOSTICS
.I Malloc, realloc
and
.I calloc
return a null pointer (0) if there is no available memory or if the arena
has been detectably corrupted by storing outside the bounds of a block.
.I Malloc
may be recompiled to check the arena very stringently on every transaction;
those sites with a source code license may check the source code to see
how this can be done.
.SH BUGS
When
.I realloc
returns 0, the block pointed to by
.I ptr
may be destroyed.
.PP
The current implementation of
.I malloc
does not always fail gracefully when system
memory limits are approached.
It may fail to allocate memory when larger free blocks could be broken
up, or when limits are exceeded because the size is rounded up.
It is optimized for sizes that are powers of two.
.PP
.I Alloca
is machine dependent; its use is discouraged.


src/machserver/src/lib/c/stdlib/malloc.c
/* 
 * malloc.c --
 *
 *	Source code for the "malloc" library procedure.  See memInt.h
 *	for overall information about how the allocator works.
 *
 * Copyright 1985, 1988 Regents of the University of California
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies.  The University of California
 * makes no representations about the suitability of this
 * software for any purpose.  It is provided "as is" without
 * express or implied warranty.
 */

#ifndef lint
static char rcsid[] = "$Header: /user5/kupfer/spriteserver/src/lib/c/stdlib/RCS/malloc.c,v 1.2 91/12/12 21:56:21 kupfer Exp $ SPRITE (Berkeley)";
#endif not lint

#include "stdlib.h"
#include "memInt.h"


/*
 * ----------------------------------------------------------------------------
 *
 * malloc --
 *
 *	This procedure allocates a chunk of memory.
 *
 * Results:
 *	The return value is a pointer to an area of at least bytesNeeded
 *	bytes of free storage.  If no memory is available, then this
 *	procedure will never return:  the MemChunkAlloc procedure 
 *	determines what will happen.
 *
 * Side effects:
 *	The returned block is marked as allocated and will not be
 *	allocated to anyone else until it is returned with a call
 *	to free.
 *
 * ----------------------------------------------------------------------------
 */

ENTRY _VoidPtr
malloc(bytesNeeded)
    register unsigned bytesNeeded;	/* How many bytes to allocate. */
{
    int			thisBlockSize, admin;
    register Address	result;
    Address		newRegion;
    int			regionSize;
    int			origSize;
    register int	index;

    LOCK_MONITOR;

#ifdef MEM_TRACE
    mem_NumAllocs++;
#endif

    if (!memInitialized) {
	MemInit();
    }

    origSize = bytesNeeded;

    /* 
     * Handle binned objects quickly, if possible.
     */

    bytesNeeded = BYTES_TO_BLOCKSIZE(bytesNeeded);
    index = BLOCKSIZE_TO_INDEX(bytesNeeded);
    if (bytesNeeded <= BIN_SIZE) {
	result = memFreeLists[index];
	if (result == NOBIN) {
	    goto largeBlockAllocator;
	}
	if (result == (Address) NULL) {

	    /*
	     * There aren't currently any free objects of this size.
	     * Call the client's function to obtain another region
	     * from the system.  While we're at it, get a whole bunch
	     * of objects of this size once and put all but one back
	     * on the free list.
	     */

	    regionSize = BLOCKS_AT_ONCE * bytesNeeded;
	    newRegion = MemChunkAlloc(regionSize);

	    while (regionSize >= bytesNeeded) {
		SET_ADMIN(newRegion, memFreeLists[index]);
		memFreeLists[index] = newRegion;
		mem_NumBlocks[index] += 1;
		newRegion += bytesNeeded;
		regionSize -= bytesNeeded;
	    }
	    result = memFreeLists[index];
	}

	memFreeLists[index] = (Address) GET_ADMIN(result);
	SET_ADMIN(result, MARK_DUMMY(bytesNeeded));

#ifdef MEM_TRACE
	mem_NumBinnedAllocs[index] += 1;
	SET_PC(result);
	SET_ORIG_SIZE(result, origSize);
	MemDoTrace(TRUE, result, (Address)NULL, bytesNeeded);
#endif MEM_TRACE

	UNLOCK_MONITOR;
	return((_VoidPtr)(result+sizeof(AdminInfo)));
    }

    /*
     * This is a large object.  Step circularly through the blocks
     * in the list, looking for one that's large enough to hold
     * what's needed.
     */

largeBlockAllocator:
    if (origSize == 0) {
	return (_VoidPtr) NULL;
    }
#ifdef MEM_TRACE
    mem_NumLargeAllocs += 1;
#endif
    result = memCurrentPtr;
    while (TRUE) {
	Address nextPtr;

#ifdef MEM_TRACE
	mem_NumLargeLoops += 1;
#endif
	admin = GET_ADMIN(result);
	thisBlockSize = SIZE(admin);
	nextPtr = result+thisBlockSize;
	if (!IS_IN_USE(admin)) {
	
	    /*
	     * Several blocks in a row could have been freed since the last
	     * time we were here.  If so, merge them together.
	     */

	    while (!IS_IN_USE(GET_ADMIN(nextPtr))) {
		thisBlockSize += SIZE(GET_ADMIN(nextPtr));
		admin = MARK_FREE(thisBlockSize);
		SET_ADMIN(result, admin);
		nextPtr = result+thisBlockSize;
	    }
	    if (thisBlockSize >= bytesNeeded) {
		break;
	    }
	    if (thisBlockSize > memLargestFree) {
		memLargestFree = thisBlockSize;
	    }
	}

	/*
	 * This block won't do the job.  Go on to the next block.
	 */

	if (nextPtr != memLast) {
	    result = nextPtr;
	    continue;
	}

	/*
	 * End of the list.  If there's any chance that there's
	 * enough free storage in the list to satisfy the request,
	 * then go back to the beginning of the list and start
	 * again.
	 */

	if ((memLargestFree + memBytesFreed) > bytesNeeded) {
	    memLargestFree = 0;
	    memBytesFreed = 0;
	    result = memFirst;
	    continue;
	}

	/*
	 * Not enough free space in the list.  Allocate a new region
	 * of memory and add it to the list.
	 */

	if (bytesNeeded < MIN_REGION_SIZE) {
	    regionSize = MIN_REGION_SIZE;
	} else {
	    regionSize = bytesNeeded + sizeof(AdminInfo);
	}
	newRegion = MemChunkAlloc(regionSize);
	mem_NumLargeBytes += regionSize;

	/*
	 * If the new region immediately follows the end of the previous
	 * region, merge them together.  At this point result always
	 * points to a block of memory adjacent to and preceding the block
	 * of memory pointed to by memLast (memLast == nextPtr).  Thus it
	 * may be possible to merge the new region with both the region
	 * at result and the one at nextPtr.
	 */

	if (newRegion == (nextPtr+sizeof(AdminInfo))) {
	    newRegion = nextPtr;
	    regionSize += sizeof(AdminInfo);
	    if (!IS_IN_USE(admin)) {
		newRegion = result;
		regionSize += SIZE(admin);
	    }
	} else {
	    SET_ADMIN(nextPtr, MARK_DUMMY(newRegion - nextPtr));
	}

	/*
	 * Create a dummy block at the end of the new region, which links
	 * to "last".
	 */
	
	SET_ADMIN(newRegion, MARK_FREE(regionSize - sizeof(AdminInfo)));
	memLast = newRegion + regionSize - sizeof(AdminInfo);
	SET_ADMIN(memLast, MARK_DUMMY(0));

	/*
	 * Continue scanning the list (try result again in case it
	 * merged with the new region).
	 */
    }

    /*
     * At this point we've got a block that's large enough.  If it's
     * larger than needed for the object, put the rest back on the
     * free list (note: even if the remnant is smaller than the smallest
     * large object, which means it'll be used by itself, we put it back
     * on the list so it can merge with either this block or the next,
     * whichever gets freed first).
     */

    if (thisBlockSize < (bytesNeeded + sizeof(AdminInfo))) {
	SET_ADMIN(result, MARK_IN_USE(admin));
    } else {
	SET_ADMIN(result+bytesNeeded, MARK_FREE(thisBlockSize-bytesNeeded));
	SET_ADMIN(result, MARK_IN_USE(bytesNeeded));
    }

#ifdef MEM_TRACE
    SET_PC(result);
    SET_ORIG_SIZE(result, origSize);
    MemDoTrace(TRUE, result, (Address) NULL, bytesNeeded);
#endif MEM_TRACE
    memCurrentPtr = result;

    UNLOCK_MONITOR;
    return((_VoidPtr)(result+sizeof(AdminInfo)));
}
